### PR TITLE
Rootless Podman security update

### LINKED ISSUE #2377: Security concerns using podman in rootless mode
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

Podman creates open postfix relay

### What happened and when does this occur?

```Markdown
As you state in your [mailserver.env](https://github.com/docker-mailserver/docker-mailserver/blob/b2b701a4786f91fe266cca4d077a088345fdff5d/mailserver.env#L53), the `network` option might open up an open relay when specifying the wrong network. The default is `localhost`.

When running Podman in rootless mode, all source IP's get translated to localhost by default, so the relay is publicly accessible to the whole WAN side and there is no mention/warning about this anywhere in the wiki.
```

### What did you expect to happen?

```Markdown
I expected this to be an `out-of-the-box secure` solution. I did not know about Podmans behaviour earlier and I think users should be warned or ideally the default config should be updated to secure Podman in rootless mode.
```

### How do we replicate the issue?

```Markdown
1. Launch DMS with Podman in rootless mode
2. Expose port 25
3. The gateway is now open to the public -> Watch your mailserver get used as a spam-distributor :)
```

### DMS version

v10.4.0

### What operating system is DMS running on?

Linux

### What instruction set architecture is DMS running on?

x86_64 / AMD64

### What container orchestration tool are you using?

Podman (not officially supported)

### docker-compose.yml

```yaml
It's the default config provided by the master branch
```

### Relevant log output

_No response_

### Other relevant information

```Markdown
This was tested on AlmaLinux 8.5 using Podman 3.3.1 in rootless mode with podman-compose v1.0.2.
Unprivileged port bindings start at 25, I have not configured firewalld redirects to unprivileged ports (but I guess this has no impact on the issue?)
```

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am inexperienced with mail servers
- [ ] I am uncomfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

### Improvements to this form?

_No response_

### PR DESCRIPTION
**Pull request by p-fruck**
# Description

This PR includes security related documentation to run docker-mailserver with podman in rootless mode. See the linked issue for more details on the security concerns.

Personally I would like to set `PERMIT_DOCKER=none` to the new default in `mailserver.env` because it enforces authentication even from localhost and therefore is more secure, but this would result in a breaking change for users that have authentication-less setups for their localhost. The decision is up to you :)

Fixes #2377

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by georglauterbach** [comment]
I recently merged a PR that changed the README. You'll need to resolve the (small) conflict there. I'd also like you to relocate the information in the documentation, at the end of the `### Create a docker-compose Environment` section, above the `### Get up and running` heading. I think this place is more fitting.

**Comment by p-fruck** [comment]
I've rebased to the latest master branch, missed the last commit :sweat_smile:

**Comment by casperklein** [comment]
Not much time right now, so just two quick thoughts:

1. This could break the update-check https://github.com/docker-mailserver/docker-mailserver/blob/4cb617f7f90bc8716601f2a07cbae8a001da30e0/target/scripts/update-check.sh#L46
2. You might want to add a test for this in https://github.com/docker-mailserver/docker-mailserver/blob/master/test/permit_docker.bats

File: target/scripts/startup/setup-stack.sh [hunk 1]
```diff
@@ -1154,6 +1154,11 @@ function _setup_docker_permit
   done < <(ip -o -4 addr show type veth | grep -E -o '[0-9\.]+/[0-9]+')
 
   case "${PERMIT_DOCKER}" in
+    "none" )
+      _notify 'inf' "Clearing my networks"
```
Comment (georglauterbach) [inline target/scripts/startup/setup-stack.sh:1158]: I'd rather use `_notify 'inf' "Clearing Postfix's 'mynetworks'"

**Review by georglauterbach — COMMENTED** [review]
Other than the one comment, @casperklein mentioned the update check. If these two concerns are resolved, I think this PR is fine :D

**Comment by p-fruck** [comment]
I've applied the suggestion and added a simply testcase. @casperklein you mentioned the issue with the updatescript, how should we overcome that? Authenticate against the server using credentials supplied as environment variables, or simply output to the terminal and leave it up to the user to handle the output if `PERMIT_DOCKER` is explicitly set to `none`?

File: mailserver.env [hunk 2]
```diff
@@ -46,9 +46,12 @@ UPDATE_CHECK_INTERVAL=1d
 # **WARNING**: Adding the docker network's gateway to the list of trusted hosts, e.g. using the `network` or
 # `connected-networks` option, can create an open relay
 # https://github.com/docker-mailserver/docker-mailserver/issues/1405#issuecomment-590106498
-# empty => localhost only
-# host => Add docker host (ipv4 only)
-# network => Add all docker containers (ipv4 only)
+# The same can happen for rootless podman. To prevent this, set the value to "none" or configure slirp4netns
+# https://github.com/docker-mailserver/docker-mailserver/issues/2377
+# <empty> => localhost and container ip only
```
Comment (casperklein) [inline mailserver.env:51]:
```suggestion
# <empty> => container ip only
```

**Comment by casperklein** [comment]
I revoked my comment about the update-check (see my edited comment above). But to be sure, you can test it with something like this:

```bash
docker exec -it mailserver bash -c 'echo "10.0.0" > /VERSION && supervisorctl restart update-check'
```

**Comment by p-fruck** [comment]
I see, just did a quick test and it works as you described, thanks :+1:

**Review by casperklein — APPROVED** [review]
Excellent first contribution 👌

**Review by wernerfred — APPROVED** [review]
LGTM, thanks

**Review by georglauterbach — APPROVED** [review]

**Comment by casperklein** [comment]
Two more things that came to my mind are `LOGWATCH` and `PFLOGSUMM`. Do you have a config like that and can check if you still get those mails daily?

```
- LOGWATCH_INTERVAL=daily
- LOGWATCH_RECIPIENT=foo@example.com
- PFLOGSUMM_TRIGGER=daily_cron
- PFLOGSUMM_RECIPIENT=foo@example.com
```

**Comment by p-fruck** [comment]
I don't have at the moment, but I could set it up within the next few days if you like

**Comment by casperklein** [comment]
That would be great.

**Comment by casperklein** [comment]
@docker-mailserver/maintainers Is there a way to detect within the container, if it's run by podman? If that is possible, we should print a open-relay warning, if `PERMIT_DOCKER` is empty.

We could also make `PERMIT_DOCKER=none` as default in the future to provide a more secure out-of-the-box solution, despite podman isn't officially supported. But then `LOGWATCH` and `PFLOGSUMM` needs to be adjusted to use authentication (if neccesary, let's wait for p-frucks feedback).

**Comment by wernerfred** [comment]
Just did a short Research and found the following:

The best check If the runtime being Docker or Podman would be to look for /run/.containerenv (Podman) vs /run/.dockerenv (Docker)

Haven't tested it myself though

**Comment by polarathene** [comment]
> .dockerenv

That is old advice IIRC, I remember looking into detecting if operating in a Docker container a while back and that approach no longer works (it was never officially documented by Docker either) as the file is no longer present AFAIK.

The easiest approach of course would be to just add a `podman` prefix/suffix tag variant that sets an ENV to detect, or very early on in README/docs explictly direct Podman users to a relevant docs page that details anything they should be aware of (might be missed if following third-party guides elsewhere though).

Changing defaults is probably fine too, but should wait until a major release right?

**Comment by casperklein** [comment]
>  as the file is no longer present AFAIK.

confirm 😞

**Comment by polarathene** [comment]
There was some other approaches I recall being advised, one regarding cgroups but it was flakey IIRC (_there's v1 and v2, along with differences between container runtimes and possibly kernel versions I think.._) and was specifically for detecting running in a containerized environment.

If Podman does still make that file available, that's all you'd really need I guess to pull off the detection, but if it's not officially documented anywhere, it might disappear one day like the Docker one did. We could also just add a `PODMAN=1` ENV or similar for users running podman to use, but that seems kinda redundant since awareness of our docs for podman (_with `PERMIT_DOCKER` warning added_) would be just as effective.

**Comment by p-fruck** [comment]
In case you want to consider a rather hacky solution, you might give this command a try
```sh
if grep 'host.containers.internal' /etc/hosts > /dev/null; then echo podman; else echo docker; fi
```

Also I just confirmend @polarathene's guess that podman (tested with `v3.4.4`) still creates the `/run/.containerenv` file

**Comment by polarathene** [comment]
When running with host mode networking, at least with Docker it will use the host system `/etc/hosts` instead IIRC, so that approach may not be reliable for some users (maybe podman handles that differently though?).

**Comment by p-fruck** [comment]
You are right, running podman in host mode does not add the `host.containers.internal` entry to the hosts file. On the other hand I'm not sure if the IP rebinding issues exist when running the container in host mode, because from what I understood the IP rebinding happens in podmans rootless overlay network, which is unused in host mode. Either way the `/run/.containerenv` might be the more reliable option

**Comment by p-fruck** [comment]
Another think that came to my mind during the setup of my new mailserver: Do you think it's reliable to check for the present network interfaces? That way also the default network interface could be dynamically changed and podman users would not have to specify `NETWORK_INTERFACE=tap0` in their environment file.

Regarding the `LOGWATCH` and `PFLOGSUMM` checks: I need to wait for my current email provider to update their DNS cache so I can send mails to my other address :unamused: How would you test such a setup? Do you have two local mail servers that communicate with each other?

**Comment by casperklein** [comment]
I am not sure if I understand you correctly. Why would you need a second DMS instance?

Another podman question: Can you confirm, that [this](https://github.com/docker-mailserver/docker-mailserver/issues/2411#issuecomment-1043299434) solution does also prevent an open-relay?
That might be a solution, if `LOGWATCH` doesn't work anymore with ``PERMIT_DOCKER=none` set.

**Comment by p-fruck** [comment]
> people not reading the documentation properly

One of them being me, I guess :sweat_smile: 

> Why would you need a second DMS instance?

In case I wanted to test inter server communication. But you are right, it's not required for the current usecase

> Can you confirm, that https://github.com/docker-mailserver/docker-mailserver/issues/2411#issuecomment-1043299434 solution does also prevent an open-relay

I can, but I didn't expect the result. Using the [slirp4netns](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/podman/#use-the-slip4netns-network-driver) approach (compatible with v10.4.0 and older) I get a  client host rejected message:
```
HELO example.com
250 mail.example.com
mail from: admin@example.com          
250 2.1.0 Ok
rcpt to: a.b@c.de
554 5.7.1 <mail.example.com[10.88.2.2]>: Client host rejected: Access denied
```

The solution shown in die issue produces a relay access denied error:
```
HELO example.com
220 mail.example.com ESMTP
250 mail.example.com
mail from: admin@example.com
250 2.1.0 Ok
rcpt to: a.b@c.de
554 5.7.1 <a.b@c.de>: Relay access denied
```

Not what I've expected, but both solutions seem to work

**Comment by casperklein** [comment]
> We could also make PERMIT_DOCKER=none as default in the future to provide a more secure out-of-the-box solution, despite podman isn't officially supported. But then LOGWATCH and PFLOGSUMM needs to be adjusted to use authentication (if neccesary, let's wait for p-frucks feedback).

I did my own tests with `PERMIT_DOCKER=none` and the mails from `LOGWATCH & PFLOGSUMM` were send successful.

**Pull request merged**